modules/rules_nodejs/4.4.6/patches/extensions.patch

```diff
@@ -0,0 +1,62 @@
+commit a3e63c15436edd1eac2646470034ac7f8a91dc2c
+Author: Alex Eagle <alex@aspect.dev>
+Date:   Sat Dec 11 15:12:06 2021 -0800
+
+    feat: allow usage from bzlmod
+    
+    Fixes #3138
+
+diff --git a/nodejs/extensions.bzl b/nodejs/extensions.bzl
+new file mode 100644
+index 00000000..9b57598e
+--- /dev/null
++++ b/nodejs/extensions.bzl
+@@ -0,0 +1,25 @@
++"extensions for bzlmod"
++
++load(":repositories.bzl", "nodejs_register_toolchains")
++
++node_toolchain = tag_class(attrs = {
++    "name": attr.string(doc = "Base name for generated repositories"),
++    "node_version": attr.string(doc = "Version of the Node.js interpreter"),
++})
++
++def _toolchain_extension(module_ctx):
++    # according to docs, modules iteration order is guaranteed to
++    # be the same as breadth-first search starting from the root module.
++    for mod in module_ctx.modules:
++        if len(mod.tags.toolchain) != 1:
++            fail("Exactly one node_toolchain can be created for a given name")
++        nodejs_register_toolchains(
++            name = mod.tags.toolchain[0].name,
++            node_version = mod.tags.toolchain[0].node_version,
++            register = False,
++        )
++
++node = module_extension(
```

Comment:
Do I understand correctly that, with Bzlmod,`rules_nodejs` always requires the end user to declare the required toolchain and register it in their MODULE.bazel file? With the current code, if one of my dependency declares a node toolchain, as a end user, I won't be able to override it (because only one can be defined)

Do you think it would be better to provide a default toolchain and also allow the end user register their own. I believe the one from the root module will be selected if multiple toolchains are registered. @Wyverald can confirm that.

Reply:
Great point, I was wondering the same.

With `WORKSPACE` the first toolchain registration wins, so I assumed this would happen with bzlmod as well. I'll do a bit of testing what happens if I declare a toolchain in the rules_nodejs MODULE.bazel file, does that mean users can skip registering one, and also they can declare their own at a specific node version?

Reply:
> does that mean users can skip registering one, and also they can declare their own at a specific node version?

Yes, I think that should be the case.

Reply:
okay I made a little test fixture:

https://gist.github.com/alexeagle/4c08e4f06c6dacb3f9ea1e764b99f860

and added a toolchain here

```
diff --git a/modules/rules_nodejs/4.4.6/MODULE.bazel b/modules/rules_nodejs/4.4.6/MODULE.bazel
index 66d1c2e..4908981 100644
--- a/modules/rules_nodejs/4.4.6/MODULE.bazel
+++ b/modules/rules_nodejs/4.4.6/MODULE.bazel
@@ -7,7 +7,12 @@ module(
     name = "rules_nodejs",
     version = "4.4.6",
     compatibility_level = 1,
+    toolchains_to_register = ["@node_toolchains//:all"],
 )
 
 bazel_dep(name = "bazel_skylib", version = "1.0.3")
 bazel_dep(name = "platforms", version = "0.0.4")
+
+node = use_extension("@rules_nodejs//nodejs:extensions.bzl", "node")
+node.toolchain(name = "node", node_version = "16.13.1")
+use_repo(node, "node_toolchains")
```

What I observe:

1) the user can no longer declare their toolchain using the name "node" because the upstream module already claimed that name, not surprising I guess

```
Error in repository_rule: A repo named node_darwin_amd64 is already generated by this module extension at /private/var/tmp/_bazel_alexeagle/cc6d8a65f2c108b871f5b3d52865faf4/external/rules_nodejs.4.4.6/nodejs/repositories.bzl:732:26
```

2) The first test succeeds, where the user declares their own toolchain, great

3) I can't get it to work when the user tries to use the "node_toolchains" declared by rules_nodejs/MODULE.bazel, as you'll see from that repro

```
ERROR: /Users/alexeagle/repros/use_nodejs_bzlmod/BUILD:32:8: no such package '@node_toolchains//': The repository '@node_toolchains' could not be resolved: Repository '@node_toolchains' is not visible from repository '@' and referenced by '//:use_default_node_toolchain'
```

what am I missing in the user's MODULE.bazel to allow them to use the default one?

Reply:
the user's MODULE.bazel additionally needs to `use_repo` "node_toolchains", not just "testnode_toolchains". `use_repo` only brings a repo into the scope of the current repo.

Reply:
I'm happy to jump on a call if it's easier. I also have problems with a follow-up PR here adding other toolchains

Reply:
@Wyverald but I don't have access to the `node` variable that needs to be the first argument to `use_repo` right?

Reply:
AH! silly me, it's the same extension I'm loading myself. Yes it works, thank you both!